neo/Ledger/Blockchain.cs

```diff
@@ -356,18 +356,27 @@ private RelayResultReason OnNewTransaction(Transaction transaction)
                 return RelayResultReason.Invalid;
             if (!Plugin.CheckPolicy(transaction))
                 return RelayResultReason.Unknown;
-            mem_pool.TryAdd(transaction.Hash, transaction);
+            mem_pool.TryAdd(transaction.Hash, transaction);
             if (mem_pool.Count > MemoryPoolSize)
             {
-                UInt256[] delete = mem_pool.Values.AsParallel()
-                    .OrderBy(p => p.NetworkFee / p.Size)
-                    .ThenBy(p => p.NetworkFee)
-                    .ThenBy(p => new BigInteger(p.Hash.ToArray()))
-                    .Take(mem_pool.Count - MemoryPoolSize)
-                    .Select(p => p.Hash)
-                    .ToArray();
-                foreach (UInt256 hash in delete)
-                    mem_pool.TryRemove(hash, out _);
+                UInt256[] oldfree = mem_pool.GetValuesBy(p => p.timestamp < DateTime.UtcNow.AddSeconds(-SecondsPerBlock * 20) && p.tx.NetworkFee == Fixed8.Zero)
+                    .Select(p => p.Hash)
+                    .ToArray();
+                foreach (UInt256 hash in oldfree)
+                    mem_pool.TryRemove(hash, out _);
+
+                if (mem_pool.Count > MemoryPoolSize)
+                {
+                    UInt256[] delete = mem_pool.Values.AsParallel()

```

Comment:
this should not be happening on every transaction with a large mem_pool. This should be done in the code after persisting a block

Reply:
See my PR as an example of where this should be done: https://github.com/neo-project/neo/pull/394